machineconfiguration/v1/types.go

```diff
@@ -745,3 +749,169 @@ type ContainerRuntimeConfigList struct {
 
 	Items []ContainerRuntimeConfig `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineState describes the health of the Machines on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineStateSpec `json:"spec"`
+	// +optional
+	Status MachineStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineState `json:"items"`
+}
+
+// MachineStateSpec describes the type of State we are managing
+type MachineStateSpec struct {
+	// Kind describes the type of MachineState
+	// +optional
+	Kind   MachineStateType   `json:"machineStateKind,omitempty"`
```

Comment:
nit: Why `MachineStateType` and `machineStateKind`?  Can we pick either type or kind and unify?  E.g.

```go
Kind   MachineStateKind   `json:"machineStateKind,omitempty"`
```

or:

```go
Type MachineStateType `json:"type,omitempty"`
```

Also, what is the use-case for `optional` and `omitempty`?  It seems like we could require typing, so that consumers could rely on the field being populated (with a non-zero length requirement too?), and not have to worry about how to to mix un-typed states in among typed states when rendering.